rados/omap.go

```diff
@@ -10,55 +10,53 @@ import (
 	"unsafe"
 )
 
+const maxIdx = 1<<31 - 1 // 2GB, max int32 value, should be safe
+const ptrSize = C.size_t(unsafe.Sizeof(unsafe.Pointer(nil)))
+
 // SetOmap appends the map `pairs` to the omap `oid`
 func (ioctx *IOContext) SetOmap(oid string, pairs map[string][]byte) error {
 	c_oid := C.CString(oid)
 	defer C.free(unsafe.Pointer(c_oid))
 
-	var s C.size_t
-	var c *C.char
-	ptrSize := unsafe.Sizeof(c)
+	numOfPairs := len(pairs)
+
+	cKeysMem := C.malloc(C.size_t(numOfPairs) * ptrSize)
+	cValuesMem := C.malloc(C.size_t(numOfPairs) * ptrSize)
+	cLengthsMem := C.malloc(C.size_t(numOfPairs) * C.sizeof_size_t)
 
-	c_keys := C.malloc(C.size_t(len(pairs)) * C.size_t(ptrSize))
-	c_values := C.malloc(C.size_t(len(pairs)) * C.size_t(ptrSize))
-	c_lengths := C.malloc(C.size_t(len(pairs)) * C.size_t(unsafe.Sizeof(s)))
+	defer C.free(cKeysMem)
+	defer C.free(cValuesMem)
+	defer C.free(cLengthsMem)
 
-	defer C.free(unsafe.Pointer(c_keys))
-	defer C.free(unsafe.Pointer(c_values))
-	defer C.free(unsafe.Pointer(c_lengths))
+	cKeys := (*[maxIdx]*C.char)(cKeysMem)[:numOfPairs:numOfPairs]
```

Comment:
A comment explaining what is done here would be helpful. This hurts my brain a little.

Also, I've not seen the `[:variable:variable]` notation before. What does that do?

Reply:
Your comment let me think that it's better to encapsulate this code in a separate package. I also added an explanation there.